Automattic/wp-calypso
# Redux: Convert billingTransactions state slice to TypeScript

## Proposed Changes

This PR adds TypeScript types to the `state.billingTransactions` Redux state slice. It adds the state types to `IAppState` which can be used in selectors and converts many of the selectors, actions, and reducers to TypeScript.

This makes a subtle change to `useRedirectToHistoryPageOnWrongSiteForTransaction` because it could previously have caused a fatal error when used by the `ReceiptView` component. It's passed the result of the `getPastBillingTransaction()` Redux selector, but that selector can return either a past billing transaction or an "individual" billing transaction (a quite different object stored in a different subtree of the `billingTransactions` state). That issue was added in https://github.com/Automattic/wp-calypso/pull/46224.

## Testing Instructions

Automated tests should be sufficient, although this does replace some uses of lodash, so it's worth manually looking to see if those were done safely.

To test to make sure these changes don't break anything, visit the site-level purchases page for a site with at least one prior purchase. Click on "Billing history" (so the URL should be something like `/purchases/billing-history/example.com`). Then click on a receipt and verify that it displays as expected.
